File: hexwar/py/hexwar/scenario.py
# -*- coding: utf-8 -*-
"""시나리오 로더 — golden/scenario.txt 를 읽어 맵과 유닛을 만든다.

   도스 게임의 시나리오 파일도 대개 이런 모양이었다. 바이너리가 아니라
   텍스트인 것은 이 교재의 편의이고, 대신 '한 글자 = 한 헥스' 라는
   원본의 성질은 그대로 지킨다.
"""

import io
import os

from . import hexcoord as H
from .hexmap import HexMap, CHAR_TO_TERRAIN, MAP_W, MAP_H
from .units import UnitPool

GOLDEN = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(
    os.path.abspath(__file__)))), 'golden')


def parse(text):
    """[terrain] [elev] [road] [units] [objectives] 다섯 블록을 나눈다."""
    blocks = {}
    cur = None
    for raw in text.split('\n'):
        line = raw.rstrip('\r')
        if not line or line.startswith(';'):
            continue
        if line.startswith('[') and line.endswith(']'):
            cur = line[1:-1]
            blocks[cur] = []
            continue
        if cur is None:
            raise ValueError('블록 밖의 줄: %r' % line)
        blocks[cur].append(line)
    return blocks


def load(path=None):
    """(맵, 유닛풀, 목표헥스목록) 을 돌려준다."""
    path = path or os.path.join(GOLDEN, 'scenario.txt')
    blocks = parse(io.open(path, encoding='utf-8').read())

    terr = blocks['terrain']
    elev = blocks['elev']
    road = blocks['road']
    if not (len(terr) == len(elev) == len(road) == MAP_H):
        raise ValueError('맵 높이가 %d 이 아니다' % MAP_H)

    m = HexMap(MAP_W, MAP_H)
    for row in range(MAP_H):
        if not (len(terr[row]) == len(elev[row]) == len(road[row]) == MAP_W):
            raise ValueError('%d행의 너비가 %d 이 아니다' % (row, MAP_W))
        for col in range(MAP_W):
            t = CHAR_TO_TERRAIN[terr[row][col]]
            e = int(elev[row][col])
            rd = 1 if road[row][col] == 'R' else 0
            m.set_cell(col, row, t, e, rd)

    pool = UnitPool()
    for line in blocks['units']:
        side, kind, col, row = (int(x) for x in line.split())
        q, r = H.oddr_to_axial(col, row)
        uid = pool.spawn(side, kind, q, r)
        m.occupant[m.idx(col, row)] = uid

    objectives = []
    for line in blocks['objectives']:
        col, row = (int(x) for x in line.split())
        objectives.append(H.oddr_to_axial(col, row))

    return m, pool, objectives

File: hexwar/lua/hexwar/scenario.lua
-- 시나리오 로더 — golden/scenario.txt

local H = require('hexwar.hexcoord')
local hexmap = require('hexwar.hexmap')
local units = require('hexwar.units')

local M = {}

-- 이 파일 기준으로 golden/ 을 찾는다. lua 는 __FILE__ 이 없으므로
-- 호출자가 정해 준 경로(HEXWAR_GOLDEN)를 우선 쓰고, 없으면 상대 경로.
function M.golden_dir()
  return os.getenv('HEXWAR_GOLDEN') or '../golden'
end

function M.parse(text)
  local blocks, cur = {}, nil
  -- 루아 5.4부터 for 의 제어 변수는 상수다. 그래서 지역 변수로 한 번 받는다.
  for raw in (text .. '\n'):gmatch('([^\n]*)\n') do
    local line = (raw:gsub('\r$', ''))
    if line ~= '' and line:sub(1, 1) ~= ';' then
      local name = line:match('^%[(.+)%]$')
      if name then
        cur = name
        blocks[cur] = {}
      else
        if not cur then error('블록 밖의 줄: ' .. line) end
        local t = blocks[cur]
        t[#t + 1] = line
      end
    end
  end
  return blocks
end

function M.load(path)
  path = path or (M.golden_dir() .. '/scenario.txt')
  local f = assert(io.open(path, 'rb'))
  local blocks = M.parse(f:read('a'))
  f:close()

  local terr, elev, road = blocks.terrain, blocks.elev, blocks.road
  if #terr ~= hexmap.MAP_H or #elev ~= hexmap.MAP_H or #road ~= hexmap.MAP_H then
    error('맵 높이가 ' .. hexmap.MAP_H .. ' 이 아니다')
  end

  local m = hexmap.new()
  for row = 0, hexmap.MAP_H - 1 do
    local tl, el, rl = terr[row + 1], elev[row + 1], road[row + 1]
    if #tl ~= hexmap.MAP_W or #el ~= hexmap.MAP_W or #rl ~= hexmap.MAP_W then
      error(row .. '행의 너비가 ' .. hexmap.MAP_W .. ' 이 아니다')
    end
    for col = 0, hexmap.MAP_W - 1 do
      local t = hexmap.CHAR_TO_TERRAIN[tl:sub(col + 1, col + 1)]
      local e = tonumber(el:sub(col + 1, col + 1))
      local rd = rl:sub(col + 1, col + 1) == 'R' and 1 or 0
      m:set_cell(col, row, t, e, rd)
    end
  end

  local pool = units.new_pool()
  for _, line in ipairs(blocks.units) do
    local side, kind, col, row = line:match('(%-?%d+) (%-?%d+) (%-?%d+) (%-?%d+)')
    side, kind, col, row = tonumber(side), tonumber(kind), tonumber(col), tonumber(row)
    local q, r = H.oddr_to_axial(col, row)
    local uid = pool:spawn(side, kind, q, r)
    m.occupant[m:idx(col, row)] = uid
  end

  local objectives = {}
  for _, line in ipairs(blocks.objectives) do
    local col, row = line:match('(%-?%d+) (%-?%d+)')
    local q, r = H.oddr_to_axial(tonumber(col), tonumber(row))
    objectives[#objectives + 1] = { q, r }
  end

  return m, pool, objectives
end

return M
